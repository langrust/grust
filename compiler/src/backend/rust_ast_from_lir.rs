//! Rust AST construction from [LIR types](mod@crate::lir).

/// RustAST block construction from LIR block.
pub mod block;
/// RustAST expression construction from LIR expression.
pub mod expression;
/// RustAST item construction from LIR item.
pub mod item;
/// RustAST pattern construction from LIR pattern.
pub mod pattern;
/// RustAST project construction from LIR project.
pub mod project;
/// RustAST statement construction from LIR statement.
pub mod statement;
/// RustAST type construction from LIR type.
pub mod typ;
