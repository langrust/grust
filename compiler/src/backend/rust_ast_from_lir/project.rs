use std::{
    collections::{BTreeMap, BTreeSet},
    path::Path,
};

use crate::{
    backend::rust_ast_from_lir::item::{
        array_alias::rust_ast_from_lir as array_alias_rust_ast_from_lir,
        enumeration::rust_ast_from_lir as enumeration_rust_ast_from_lir,
        function::rust_ast_from_lir as function_rust_ast_from_lir,
        node_file::rust_ast_from_lir as node_file_rust_ast_from_lir,
        structure::rust_ast_from_lir as structure_rust_ast_from_lir,
    },
    lir::{item::Item, project::Project},
};
use itertools::Itertools;
use proc_macro2::Span;
use syn::*;

#[derive(Debug)]
/// Rust project resulting from compiling.
pub struct RustASTProject {
    /// Project's directory.
    pub directory: String,
    files: BTreeMap<String, File>,
    crates: BTreeSet<String>,
}
impl RustASTProject {
    fn new() -> Self {
        RustASTProject {
            files: Default::default(),
            directory: Default::default(),
            crates: Default::default(),
        }
    }

    fn add_file(&mut self, path: &str, file: File) {
        self.files.insert(path.to_owned(), file);
    }

    /// Set project's directory.
    pub fn set_parent<P>(&mut self, path: P)
    where
        P: AsRef<std::path::Path>,
    {
        let subdirectory = std::mem::take(&mut self.directory);
        self.directory = path
            .as_ref()
            .join(subdirectory)
            .into_os_string()
            .into_string()
            .unwrap();
    }
}

/// Transform LIR item into RustAST item.
pub fn rust_ast_from_lir(project: Project) -> RustASTProject {
    let mut rust_ast_project = RustASTProject::new();

    let mut function_file = File {
        shebang: None,
        items: Default::default(),
        attrs: vec![],
    };
    let mut typedefs_file = File {
        shebang: None,
        items: Default::default(),
        attrs: vec![],
    };

    project.items.into_iter().for_each(|item| match item {
        Item::NodeFile(node_file) => {
            let (path, rust_ast_node_file) =
                node_file_rust_ast_from_lir(node_file, &mut rust_ast_project.crates);
            rust_ast_project.add_file(&path, rust_ast_node_file)
        }
        Item::Function(function) => {
            let mut rust_ast_function =
                function_rust_ast_from_lir(function, &mut rust_ast_project.crates);
            function_file.items.append(&mut rust_ast_function);
            // remove duplicated imports between functions
            function_file.items = std::mem::take(&mut function_file.items)
                .into_iter()
                .unique()
                .collect::<Vec<_>>();
        }
        Item::Enumeration(enumeration) => {
            let rust_ast_enumeration = enumeration_rust_ast_from_lir(enumeration);
            typedefs_file
                .items
                .push(syn::Item::Enum(rust_ast_enumeration))
        }
        Item::Structure(structure) => {
            let rust_ast_structure = structure_rust_ast_from_lir(structure);
            typedefs_file
                .items
                .push(syn::Item::Struct(rust_ast_structure))
        }
        Item::ArrayAlias(array_alias) => {
            let rust_ast_array_alias = array_alias_rust_ast_from_lir(array_alias);
            typedefs_file
                .items
                .push(syn::Item::Type(rust_ast_array_alias))
        }
    });

    rust_ast_project.add_file("src/functions.rs", function_file);
    rust_ast_project.add_file("src/typedefs.rs", typedefs_file);

    let mut lib_file = File {
        shebang: None,
        items: Default::default(),
        attrs: vec![],
    };
    rust_ast_project.files.iter().for_each(|(path, _)| {
        let module_name = Path::new(&path)
            .file_stem()
            .unwrap()
            .to_str()
            .unwrap()
            .to_string();
        let module_ident = Ident::new(&module_name, Span::call_site());
        let module_decl = parse_quote! { pub mod #module_ident; };
        lib_file.items.push(syn::Item::Mod(module_decl))
    });
    rust_ast_project.add_file("src/lib.rs", lib_file);

    rust_ast_project
}
