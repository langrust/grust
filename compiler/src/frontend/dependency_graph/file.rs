use petgraph::algo::toposort;
use petgraph::graphmap::DiGraphMap;

prelude! {
    error::{Error, TerminationError},
    hir::file::File,
    symbol_table::SymbolTable,

}

impl File {
    /// Generate dependency graph for every nodes/component.
    pub fn generate_dependency_graphs(
        &mut self,
        symbol_table: &SymbolTable,
        errors: &mut Vec<Error>,
    ) -> Result<(), TerminationError> {
        let File { nodes, .. } = self;

        // initialize dictionariy for reduced graphs
        let mut nodes_reduced_graphs = HashMap::new();

        // create graph of nodes
        let mut nodes_graph = DiGraphMap::new();
        nodes
            .iter()
            .for_each(|node| node.add_node_dependencies(&mut nodes_graph));

        // sort nodes according to their dependencies
        let sorted_nodes = toposort(&nodes_graph, None).map_err(|node| {
            let error = Error::NotCausalNode {
                node: symbol_table.get_name(node.node_id()).clone(),
                location: self.location.clone(),
            };
            errors.push(error);
            TerminationError
        })?;
        nodes.sort_by(|n1, n2| {
            let index1 = sorted_nodes
                .iter()
                .position(|id| *id == n1.id)
                .expect("should be in sorted list");
            let index2 = sorted_nodes
                .iter()
                .position(|id| *id == n2.id)
                .expect("should be in sorted list");

            Ord::cmp(&index2, &index1)
        });

        // ordered nodes complete their dependency graphs
        nodes
            .iter_mut()
            .map(|node| node.compute_dependencies(symbol_table, &mut nodes_reduced_graphs, errors))
            .collect::<Vec<Result<(), TerminationError>>>()
            .into_iter()
            .collect::<Result<(), TerminationError>>()?;

        Ok(())
    }
}
