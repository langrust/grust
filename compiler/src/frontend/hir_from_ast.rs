prelude! {}

/// HIR Component construction from AST Component
pub mod component;
/// HIR Contract construction from AST Contract
pub mod contract;
/// HIR Equation construction from AST Equation
pub mod equation;
/// HIR Expression construction from AST Expression
pub mod expression;
/// HIR File construction from AST File
pub mod file;
/// HIR Function construction from AST Function
pub mod function;
/// HIR Interface construction from AST Interface.
pub mod interface;
/// HIR Pattern construction from AST Pattern
pub mod pattern;
/// HIR Statement construction from AST Statement
pub mod statement;
/// HIR StreamExpression construction from AST StreamExpression
pub mod stream_expression;
pub mod typ;
/// HIR Typedef construction from AST Typedef.
pub mod typedef;

/// AST transformation into HIR.
pub trait HIRFromAST<Ctxt> {
    /// Corresponding HIR construct.
    type HIR;

    /// Transforms AST into HIR and check identifiers good use.
    fn hir_from_ast(self, ctxt: &mut Ctxt) -> TRes<Self::HIR>;
}

pub struct SimpleCtxt<'a> {
    pub syms: &'a mut SymbolTable,
    pub errors: &'a mut Vec<Error>,
}
pub struct LocCtxt<'a> {
    pub loc: &'a Location,
    pub syms: &'a mut SymbolTable,
    pub errors: &'a mut Vec<Error>,
}
pub struct PatLocCtxt<'a> {
    pub pat: Option<&'a ast::stmt::Pattern>,
    pub loc: &'a Location,
    pub syms: &'a mut SymbolTable,
    pub errors: &'a mut Vec<Error>,
}
impl<'a> SimpleCtxt<'a> {
    pub fn new(syms: &'a mut SymbolTable, errors: &'a mut Vec<Error>) -> Self {
        Self { syms, errors }
    }
    pub fn add_loc<'b>(&'b mut self, loc: &'b Location) -> LocCtxt<'b> {
        LocCtxt::new(loc, self.syms, self.errors)
    }
    pub fn add_pat_loc<'b>(
        &'b mut self,
        pat: Option<&'b ast::stmt::Pattern>,
        loc: &'b Location,
    ) -> PatLocCtxt<'b> {
        PatLocCtxt::new(pat, loc, self.syms, self.errors)
    }
}
impl<'a> LocCtxt<'a> {
    pub fn new(loc: &'a Location, syms: &'a mut SymbolTable, errors: &'a mut Vec<Error>) -> Self {
        Self { loc, syms, errors }
    }
    pub fn add_pat<'b>(&'b mut self, pat: Option<&'b ast::stmt::Pattern>) -> PatLocCtxt<'b> {
        PatLocCtxt::new(pat, self.loc, self.syms, self.errors)
    }
}
impl<'a> PatLocCtxt<'a> {
    pub fn new(
        pat: Option<&'a ast::stmt::Pattern>,
        loc: &'a Location,
        syms: &'a mut SymbolTable,
        errors: &'a mut Vec<Error>,
    ) -> Self {
        Self {
            pat,
            loc,
            syms,
            errors,
        }
    }
    pub fn remove_pat<'b>(&'b mut self) -> LocCtxt<'b> {
        LocCtxt::new(self.loc, self.syms, self.errors)
    }
    pub fn set_pat(
        &mut self,
        pat: Option<&'a ast::stmt::Pattern>,
    ) -> Option<&'a ast::stmt::Pattern> {
        std::mem::replace(&mut self.pat, pat)
    }
}
