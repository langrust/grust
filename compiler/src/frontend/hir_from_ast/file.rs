prelude! {
    ast::Ast,
    hir::interface::Interface,
}

use super::HIRFromAST;

impl HIRFromAST for Ast {
    type HIR = hir::File;

    fn hir_from_ast(
        self,
        symbol_table: &mut SymbolTable,
        errors: &mut Vec<Error>,
    ) -> TRes<Self::HIR> {
        // initialize symbol table with builtin operators
        symbol_table.initialize();

        // store elements in symbol table
        self.store(symbol_table, errors)?;

        let Ast { items } = self;

        let (typedefs, functions, components, imports, exports, services) = items.into_iter().fold(
            (vec![], vec![], vec![], vec![], vec![], vec![]),
            |(
                mut typedefs,
                mut functions,
                mut components,
                mut imports,
                mut exports,
                mut services,
            ),
             item| {
                match item {
                    ast::Item::Component(component) => {
                        components.push(component.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Function(function) => {
                        functions.push(function.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Typedef(typedef) => {
                        typedefs.push(typedef.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Service(service) => {
                        services.push(service.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Import(import) => imports.push(
                        import
                            .hir_from_ast(symbol_table, errors)
                            .map(|res| (symbol_table.get_fresh_id(), res)),
                    ),
                    ast::Item::Export(export) => exports.push(
                        export
                            .hir_from_ast(symbol_table, errors)
                            .map(|res| (symbol_table.get_fresh_id(), res)),
                    ),
                    ast::Item::ComponentImport(component) => {
                        components.push(component.hir_from_ast(symbol_table, errors))
                    }
                }
                (typedefs, functions, components, imports, exports, services)
            },
        );

        let interface = Interface {
            services: services.into_iter().collect::<TRes<Vec<_>>>()?,
            imports: imports.into_iter().collect::<TRes<_>>()?,
            exports: exports.into_iter().collect::<TRes<_>>()?,
        };

        Ok(hir::File {
            typedefs: typedefs.into_iter().collect::<TRes<Vec<_>>>()?,
            functions: functions.into_iter().collect::<TRes<Vec<_>>>()?,
            components: components.into_iter().collect::<TRes<Vec<_>>>()?,
            interface,
            location: Location::default(),
        })
    }
}
