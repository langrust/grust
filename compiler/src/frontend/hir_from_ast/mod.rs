prelude! {}

/// HIR Contract construction from AST Contract
pub mod contract;
/// HIR Equation construction from AST Equation
pub mod equation;
/// HIR Expression construction from AST Expression
pub mod expression;
/// HIR File construction from AST File
pub mod file;
/// HIR Function construction from AST Function
pub mod function;
/// HIR Interface construction from AST Interface.
pub mod interface;
/// HIR Pattern construction from AST Pattern
pub mod pattern;
/// HIR Statement construction from AST Statement
pub mod statement;
/// HIR StreamExpression construction from AST StreamExpression
pub mod stream_expression;
/// HIR Typedef construction from AST Typedef.
pub mod typedef;

/// AST transformation into HIR.
pub trait HIRFromAST {
    /// Corresponding HIR construct.
    type HIR;

    /// Transforms AST into HIR and check identifiers good use.
    fn hir_from_ast(
        self,
        symbol_table: &mut SymbolTable,
        errors: &mut Vec<Error>,
    ) -> TRes<Self::HIR>;
}

impl HIRFromAST for ast::Component {
    type HIR = hir::Node;

    // precondition: node and its signals are already stored in symbol table
    // postcondition: construct HIR node and check identifiers good use
    fn hir_from_ast(
        self,
        symbol_table: &mut SymbolTable,
        errors: &mut Vec<Error>,
    ) -> TRes<Self::HIR> {
        let ast::Component {
            ident,
            contract,
            equations,
            ..
        } = self;
        let name = ident.to_string();
        let location = Location::default();
        let id = symbol_table.get_node_id(&name, false, location.clone(), errors)?;

        // create local context with all signals
        symbol_table.local();
        symbol_table.restore_context(id);
        symbol_table.enter_in_node(id);

        let statements = equations
            .into_iter()
            .map(|equation| equation.hir_from_ast(symbol_table, errors))
            .collect::<TRes<Vec<_>>>()?;
        let contract = contract.hir_from_ast(symbol_table, errors)?;

        symbol_table.leave_node();
        symbol_table.global();

        Ok(hir::Node {
            id,
            statements,
            contract,
            location,
            graph: graph::DiGraphMap::new(),
            memory: hir::Memory::new(),
        })
    }
}
