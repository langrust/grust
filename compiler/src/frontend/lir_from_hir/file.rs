use crate::{
    hir::file::File,
    lir::{item::Item, project::Project},
    symbol_table::SymbolTable,
};

use super::LIRFromHIR;

impl LIRFromHIR for File {
    type LIR = Project;

    fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR {
        let File {
            typedefs,
            functions,
            nodes,
            interface,
            ..
        } = self;

        let mut typedefs = typedefs
            .into_iter()
            .map(|typedef| typedef.lir_from_hir(symbol_table))
            .collect();
        let mut functions = functions
            .into_iter()
            .map(|function| function.lir_from_hir(symbol_table))
            .map(Item::Function)
            .collect();
        let mut state_machines = nodes
            .into_iter()
            .map(|node| node.lir_from_hir(symbol_table))
            .map(Item::StateMachine)
            .collect();
        let execution_machine = interface.lir_from_hir(symbol_table);

        let mut items = vec![];
        items.append(&mut typedefs);
        items.append(&mut functions);
        items.append(&mut state_machines);
        items.push(Item::ExecutionMachine(execution_machine));

        Project { items }
    }
}
