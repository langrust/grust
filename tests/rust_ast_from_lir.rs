use codespan_reporting::files::{Files, SimpleFiles};

use grustine::ast::file::File;
use grustine::backend::rust_ast_from_lir::project::rust_ast_from_lir;
use grustine::frontend::hir_from_ast::file::hir_from_ast;
use grustine::frontend::lir_from_hir::file::lir_from_hir;
use grustine::parser::langrust;

#[test]
fn rust_ast_from_lir_transformation_for_counter() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let counter_id = files.add(
        "counter.gr",
        std::fs::read_to_string("tests/fixture/counter.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(counter_id, &files.source(counter_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();
    file.normalize(&mut errors).unwrap();
    let project = lir_from_hir(file);

    let project = rust_ast_from_lir(project);
    insta::assert_debug_snapshot!(project);
}

#[test]
fn rust_ast_from_lir_transformation_for_blinking() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "blinking.gr",
        std::fs::read_to_string("tests/fixture/blinking.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();
    file.normalize(&mut errors).unwrap();
    let project = lir_from_hir(file);

    let project = rust_ast_from_lir(project);
    insta::assert_debug_snapshot!(project);
}

#[test]
fn rust_ast_from_lir_transformation_for_button_management() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management.gr",
        std::fs::read_to_string("tests/fixture/button_management.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();
    file.normalize(&mut errors).unwrap();
    let project = lir_from_hir(file);

    let project = rust_ast_from_lir(project);
    insta::assert_debug_snapshot!(project);
}

#[test]
fn rust_ast_from_lir_transformation_for_button_management_condition_match() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management_condition_match.gr",
        std::fs::read_to_string("tests/fixture/button_management_condition_match.gr")
            .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();
    file.normalize(&mut errors).unwrap();
    let project = lir_from_hir(file);

    let project = rust_ast_from_lir(project);
    insta::assert_debug_snapshot!(project);
}

#[test]
fn rust_ast_from_lir_transformation_for_button_management_using_function() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management_using_function.gr",
        std::fs::read_to_string("tests/fixture/button_management_using_function.gr")
            .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();
    file.normalize(&mut errors).unwrap();
    let project = lir_from_hir(file);

    let project = rust_ast_from_lir(project);
    insta::assert_debug_snapshot!(project);
}

#[test]
fn rust_ast_from_lir_transformation_for_pid() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let pid_id = files.add(
        "pid.gr",
        std::fs::read_to_string("tests/fixture/pid.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(pid_id, &files.source(pid_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();
    file.normalize(&mut errors).unwrap();
    let project = lir_from_hir(file);

    let project = rust_ast_from_lir(project);
    insta::assert_debug_snapshot!(project);
}
