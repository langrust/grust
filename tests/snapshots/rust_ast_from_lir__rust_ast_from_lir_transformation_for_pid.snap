---
source: tests/rust_ast_from_lir.rs
expression: project
---
directory: ""
files:
  - path: src/integrate_i.rs
    items:
      - Import:
          Use:
            public_visibility: false
            tree:
              Path:
                module_name: crate
                tree:
                  Path:
                    module_name: functions
                    tree:
                      Name:
                        name: min
                        alias: ~
      - Structure:
          public_visibility: true
          name: integrate_iInput
          fields:
            - public_visibility: true
              name: x
              type:
                Identifier:
                  identifier: f64
            - public_visibility: true
              name: dt
              type:
                Identifier:
                  identifier: f64
      - Structure:
          public_visibility: true
          name: integrate_iState
          fields:
            - public_visibility: false
              name: mem_prev_i
              type:
                Identifier:
                  identifier: f64
      - Implementation:
          trait_name: ~
          type_name: integrate_iState
          items:
            - AssociatedMethod:
                signature:
                  public_visibility: true
                  name: init
                  receiver: ~
                  inputs: []
                  output:
                    Identifier:
                      identifier: integrate_iState
                body:
                  statements:
                    - ExpressionLast:
                        Structure:
                          name: integrate_iState
                          fields:
                            - name: mem_prev_i
                              expression:
                                Literal:
                                  literal:
                                    Float: 0
            - AssociatedMethod:
                signature:
                  public_visibility: true
                  name: step
                  receiver:
                    reference: false
                    mutable: false
                  inputs:
                    - - input
                      - Identifier:
                          identifier: integrate_iInput
                  output:
                    Tuple:
                      elements:
                        - Identifier:
                            identifier: integrate_iState
                        - Identifier:
                            identifier: f64
                body:
                  statements:
                    - Let:
                        pattern:
                          Identifier:
                            reference: false
                            mutable: false
                            identifier: MAX_INTEGRALE
                        expression:
                          Literal:
                            literal:
                              Float: 1000
                    - Let:
                        pattern:
                          Identifier:
                            reference: false
                            mutable: false
                            identifier: prev_i
                        expression:
                          FieldAccess:
                            expression:
                              Identifier:
                                identifier: self
                            field: mem_prev_i
                    - Let:
                        pattern:
                          Identifier:
                            reference: false
                            mutable: false
                            identifier: i
                        expression:
                          FunctionCall:
                            function:
                              Identifier:
                                identifier: min
                            arguments:
                              - Binary:
                                  left:
                                    Identifier:
                                      identifier: prev_i
                                  operator: Add
                                  right:
                                    Binary:
                                      left:
                                        FieldAccess:
                                          expression:
                                            Identifier:
                                              identifier: input
                                          field: x
                                      operator: Mul
                                      right:
                                        FieldAccess:
                                          expression:
                                            Identifier:
                                              identifier: input
                                          field: dt
                              - Identifier:
                                  identifier: MAX_INTEGRALE
                    - ExpressionLast:
                        Tuple:
                          elements:
                            - Structure:
                                name: integrate_iState
                                fields:
                                  - name: mem_prev_i
                                    expression:
                                      Identifier:
                                        identifier: i
                            - Identifier:
                                identifier: i
  - path: src/derive_d.rs
    items:
      - Structure:
          public_visibility: true
          name: derive_dInput
          fields:
            - public_visibility: true
              name: x
              type:
                Identifier:
                  identifier: f64
            - public_visibility: true
              name: dt
              type:
                Identifier:
                  identifier: f64
      - Structure:
          public_visibility: true
          name: derive_dState
          fields:
            - public_visibility: false
              name: mem_prev_x
              type:
                Identifier:
                  identifier: f64
      - Implementation:
          trait_name: ~
          type_name: derive_dState
          items:
            - AssociatedMethod:
                signature:
                  public_visibility: true
                  name: init
                  receiver: ~
                  inputs: []
                  output:
                    Identifier:
                      identifier: derive_dState
                body:
                  statements:
                    - ExpressionLast:
                        Structure:
                          name: derive_dState
                          fields:
                            - name: mem_prev_x
                              expression:
                                Literal:
                                  literal:
                                    Float: 0
            - AssociatedMethod:
                signature:
                  public_visibility: true
                  name: step
                  receiver:
                    reference: false
                    mutable: false
                  inputs:
                    - - input
                      - Identifier:
                          identifier: derive_dInput
                  output:
                    Tuple:
                      elements:
                        - Identifier:
                            identifier: derive_dState
                        - Identifier:
                            identifier: f64
                body:
                  statements:
                    - Let:
                        pattern:
                          Identifier:
                            reference: false
                            mutable: false
                            identifier: prev_x
                        expression:
                          FieldAccess:
                            expression:
                              Identifier:
                                identifier: self
                            field: mem_prev_x
                    - Let:
                        pattern:
                          Identifier:
                            reference: false
                            mutable: false
                            identifier: d
                        expression:
                          Binary:
                            left:
                              Unary:
                                operator: Brackets
                                expression:
                                  Binary:
                                    left:
                                      FieldAccess:
                                        expression:
                                          Identifier:
                                            identifier: input
                                        field: x
                                    operator: Sub
                                    right:
                                      Identifier:
                                        identifier: prev_x
                            operator: Div
                            right:
                              FieldAccess:
                                expression:
                                  Identifier:
                                    identifier: input
                                field: dt
                    - ExpressionLast:
                        Tuple:
                          elements:
                            - Structure:
                                name: derive_dState
                                fields:
                                  - name: mem_prev_x
                                    expression:
                                      FieldAccess:
                                        expression:
                                          Identifier:
                                            identifier: input
                                        field: x
                            - Identifier:
                                identifier: d
  - path: src/pid_u.rs
    items:
      - Import:
          Use:
            public_visibility: false
            tree:
              Path:
                module_name: crate
                tree:
                  Path:
                    module_name: derive_d
                    tree: Star
      - Import:
          Use:
            public_visibility: false
            tree:
              Path:
                module_name: crate
                tree:
                  Path:
                    module_name: integrate_i
                    tree: Star
      - Import:
          Use:
            public_visibility: false
            tree:
              Path:
                module_name: crate
                tree:
                  Path:
                    module_name: typedefs
                    tree:
                      Name:
                        name: GainPID
                        alias: ~
      - Structure:
          public_visibility: true
          name: pid_uInput
          fields:
            - public_visibility: true
              name: v_c
              type:
                Identifier:
                  identifier: f64
            - public_visibility: true
              name: v
              type:
                Identifier:
                  identifier: f64
            - public_visibility: true
              name: dt
              type:
                Identifier:
                  identifier: f64
      - Structure:
          public_visibility: true
          name: pid_uState
          fields:
            - public_visibility: false
              name: derive_d_e_d
              type:
                Identifier:
                  identifier: derive_dState
            - public_visibility: false
              name: integrate_i_e_i
              type:
                Identifier:
                  identifier: integrate_iState
      - Implementation:
          trait_name: ~
          type_name: pid_uState
          items:
            - AssociatedMethod:
                signature:
                  public_visibility: true
                  name: init
                  receiver: ~
                  inputs: []
                  output:
                    Identifier:
                      identifier: pid_uState
                body:
                  statements:
                    - ExpressionLast:
                        Structure:
                          name: pid_uState
                          fields:
                            - name: derive_d_e_d
                              expression:
                                FunctionCall:
                                  function:
                                    Identifier:
                                      identifier: "derive_dState::init"
                                  arguments: []
                            - name: integrate_i_e_i
                              expression:
                                FunctionCall:
                                  function:
                                    Identifier:
                                      identifier: "integrate_iState::init"
                                  arguments: []
            - AssociatedMethod:
                signature:
                  public_visibility: true
                  name: step
                  receiver:
                    reference: false
                    mutable: false
                  inputs:
                    - - input
                      - Identifier:
                          identifier: pid_uInput
                  output:
                    Tuple:
                      elements:
                        - Identifier:
                            identifier: pid_uState
                        - Identifier:
                            identifier: f64
                body:
                  statements:
                    - Let:
                        pattern:
                          Identifier:
                            reference: false
                            mutable: false
                            identifier: e
                        expression:
                          Binary:
                            left:
                              FieldAccess:
                                expression:
                                  Identifier:
                                    identifier: input
                                field: v_c
                            operator: Sub
                            right:
                              FieldAccess:
                                expression:
                                  Identifier:
                                    identifier: input
                                field: v
                    - Let:
                        pattern:
                          Tuple:
                            elements:
                              - Identifier:
                                  reference: false
                                  mutable: false
                                  identifier: derive_d_e_d
                              - Identifier:
                                  reference: false
                                  mutable: false
                                  identifier: e_d
                        expression:
                          MethodCall:
                            receiver:
                              FieldAccess:
                                expression:
                                  Identifier:
                                    identifier: self
                                field: derive_d_e_d
                            method: step
                            arguments:
                              - Structure:
                                  name: derive_dInput
                                  fields:
                                    - name: x
                                      expression:
                                        Identifier:
                                          identifier: e
                                    - name: dt
                                      expression:
                                        FieldAccess:
                                          expression:
                                            Identifier:
                                              identifier: input
                                          field: dt
                    - Let:
                        pattern:
                          Tuple:
                            elements:
                              - Identifier:
                                  reference: false
                                  mutable: false
                                  identifier: integrate_i_e_i
                              - Identifier:
                                  reference: false
                                  mutable: false
                                  identifier: e_i
                        expression:
                          MethodCall:
                            receiver:
                              FieldAccess:
                                expression:
                                  Identifier:
                                    identifier: self
                                field: integrate_i_e_i
                            method: step
                            arguments:
                              - Structure:
                                  name: integrate_iInput
                                  fields:
                                    - name: x
                                      expression:
                                        Identifier:
                                          identifier: e
                                    - name: dt
                                      expression:
                                        FieldAccess:
                                          expression:
                                            Identifier:
                                              identifier: input
                                          field: dt
                    - Let:
                        pattern:
                          Identifier:
                            reference: false
                            mutable: false
                            identifier: gain
                        expression:
                          Structure:
                            name: GainPID
                            fields:
                              - name: k_p
                                expression:
                                  Literal:
                                    literal:
                                      Float: 0.2
                              - name: k_i
                                expression:
                                  Literal:
                                    literal:
                                      Float: 1.5
                              - name: k_d
                                expression:
                                  Literal:
                                    literal:
                                      Float: 6
                    - Let:
                        pattern:
                          Identifier:
                            reference: false
                            mutable: false
                            identifier: u
                        expression:
                          Match:
                            matched:
                              Identifier:
                                identifier: gain
                            arms:
                              - pattern:
                                  Structure:
                                    name: GainPID
                                    fields:
                                      - name: k_p
                                        pattern:
                                          Identifier:
                                            reference: false
                                            mutable: false
                                            identifier: k_p
                                      - name: k_i
                                        pattern:
                                          Identifier:
                                            reference: false
                                            mutable: false
                                            identifier: k_i
                                      - name: k_d
                                        pattern:
                                          Identifier:
                                            reference: false
                                            mutable: false
                                            identifier: k_d
                                    dots: false
                                guard: ~
                                body:
                                  Binary:
                                    left:
                                      Binary:
                                        left:
                                          Binary:
                                            left:
                                              Identifier:
                                                identifier: k_p
                                            operator: Mul
                                            right:
                                              Identifier:
                                                identifier: e
                                        operator: Add
                                        right:
                                          Binary:
                                            left:
                                              Identifier:
                                                identifier: k_i
                                            operator: Mul
                                            right:
                                              Identifier:
                                                identifier: e_i
                                    operator: Add
                                    right:
                                      Binary:
                                        left:
                                          Identifier:
                                            identifier: k_d
                                        operator: Mul
                                        right:
                                          Identifier:
                                            identifier: e_d
                    - ExpressionLast:
                        Tuple:
                          elements:
                            - Structure:
                                name: pid_uState
                                fields:
                                  - name: derive_d_e_d
                                    expression:
                                      Identifier:
                                        identifier: derive_d_e_d
                                  - name: integrate_i_e_i
                                    expression:
                                      Identifier:
                                        identifier: integrate_i_e_i
                            - Identifier:
                                identifier: u
  - path: src/functions.rs
    items:
      - Function:
          signature:
            public_visibility: true
            name: min
            receiver: ~
            inputs:
              - - a
                - Identifier:
                    identifier: f64
              - - b
                - Identifier:
                    identifier: f64
            output:
              Identifier:
                identifier: f64
          body:
            statements:
              - Let:
                  pattern:
                    Identifier:
                      reference: false
                      mutable: false
                      identifier: test
                  expression:
                    Binary:
                      left:
                        Identifier:
                          identifier: a
                      operator: Grt
                      right:
                        Identifier:
                          identifier: b
              - Let:
                  pattern:
                    Identifier:
                      reference: false
                      mutable: false
                      identifier: result
                  expression:
                    IfThenElse:
                      condition:
                        Identifier:
                          identifier: test
                      then_branch:
                        statements:
                          - ExpressionLast:
                              Identifier:
                                identifier: b
                      else_branch:
                        statements:
                          - ExpressionLast:
                              Identifier:
                                identifier: a
              - ExpressionLast:
                  Identifier:
                    identifier: result
  - path: src/typedefs.rs
    items:
      - Structure:
          public_visibility: true
          name: GainPID
          fields:
            - public_visibility: true
              name: k_p
              type:
                Identifier:
                  identifier: f64
            - public_visibility: true
              name: k_i
              type:
                Identifier:
                  identifier: f64
            - public_visibility: true
              name: k_d
              type:
                Identifier:
                  identifier: f64
  - path: src/lib.rs
    items:
      - Import:
          Module:
            public_visibility: true
            name: integrate_i
      - Import:
          Module:
            public_visibility: true
            name: derive_d
      - Import:
          Module:
            public_visibility: true
            name: pid_u
      - Import:
          Module:
            public_visibility: true
            name: functions
      - Import:
          Module:
            public_visibility: true
            name: typedefs

